src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -0,0 +1,127 @@
+ * xgboost based decision trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+#define MLPACK_METHODS_XGBOOST_LOSS_FUNCTIONS_SSE_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ensemble {
+
+/**
+ * The SSE (Sum of Squared Errors) loss is a loss function to measure the
+ * quality of prediction of response values present in the node of each
+ * xgboost tree. It is also a good measure to compare the spread of two
+ * distributions. We will try to minimize this value while training.
+ *
+ * Loss = 1 / 2 * (Observed - Predicted)^2
+ */
+class SSELoss
+{
+ public:
+  /**
+   * Returns the initial predition for gradient boosting.
+   */
+  template<typename VecType>
+  typename VecType::elem_type InitialPrediction(const VecType& values)
+  {
+    return arma::accu(values) / (typename VecType::elem_type) values.n_elem;
+  }
+
+  /**
+   * Returns the first order gradient of the loss function with respect to the
+   * values.
+   *
+   * This is primarily used in calculating the residuals and split gain for the
+   * gradient boosted trees.
+   *
+   * @tparam T The type of input data. This can be both a vector or a scalar.
+   * @param observed The true observed values.
+   * @param values The values with respect to which the gradient will be
+   *     calculated.
+   */
+  template<typename T>
+  T Gradients(const T& observed, const T& values)
+  {
+    return - (observed - values);
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for scalars.
+   */
+  template<typename T>
+  T Hessians(const T& /* observed */, const T& /* values */)
+  {
+    return (T) 1;
+  }
+
+  /**
+   * Returns the second order gradient of the loss function with respect to the
+   * values. This is used only for vectors.
+   */
+  template<typename VecType,
+           class = std::enable_if_t<arma::is_Col<VecType>::value ||
+               arma::is_Row<VecType>::value>>
+  VecType Hessians(const VecType& /* observed */, const VecType& values)
+  {
+    VecType h(values.n_elem, arma::fill::ones);
+    return h;
+  }
+
+  /**
+   * Returns the pseudo residuals of the predictions.
+   * This is equal to the negative gradient of the loss function with respect
+   * to the predicted values f.
+   *
+   * @param observed The true observed values.
+   * @param f The prediction at the current step of boosting.
+   */
+  template<typename VecType>
+  VecType Residuals(const VecType& observed, const VecType& f)
+  {
+    return - Gradients(observed, f);
+  }
+
+  /**
+   * Returns the output value for the leaf in the tree.
```

Comment:
This computes the prediction value for a leaf in a regression tree, right?  Do we use a majority vote for classification, or some other rule?

Reply:
In the case of XGBoost, regardless of whether we are doing classification or regression, the weak learner is a regression tree. The majority vote kind of thing is done in the gradient boosting machine.

Take the example of binary classification. The weak learner decision tree predicts the probability of a point belonging to class 1. Then the XGBoost class will convert that probability into a class label. Similarly, for multi-class classification, we have to use one-v-one or one-v-rest strategies.

Reply:
Ok, I guess in this case, I am not really sure where this function will be used.  Is it the objective value that we are minimizing, to be computed after each iteration to see if we have converged?